aten/src/ATen/native/cpu/NaiveConvolutionTranspose3d.cpp

```diff
@@ -0,0 +1,563 @@
+          const int64_t n = input_depth * input_height * input_width;
+          const int64_t k =
+              weight.size(1) * weight.size(2) * weight.size(3) * weight.size(4);
+
+          // Do GEMM (note: this is a bit confusing because gemm assumes
+          // column-major matrices)
+          auto gemm_in_ptr = need_columns ? grad_columns.data_ptr<scalar_t>()
+              : grad_output_n.data_ptr<scalar_t>();
+          cpublas::gemm(
+              TransposeType::NoTranspose,
+              TransposeType::NoTranspose,
+              n,
+              m,
+              k,
+              1,
+              gemm_in_ptr,
+              n,
+              weight.data_ptr<scalar_t>(),
+              k,
+              0,
+              grad_input_n.data_ptr<scalar_t>(),
+              n);
+        }
+
+        // Resize output
+        if (is_batch) {
+          grad_output.resize_(
+              {n_output_plane, output_depth, output_height, output_width});
+          input.resize_(
+              {n_input_plane, input_depth, input_height, input_width});
+          grad_input.resize_(
+              {n_input_plane, input_depth, input_height, input_width});
+        }
+      });
+}
+
+void slow_conv_transpose3d_acc_grad_parameters_cpu(
+    const Tensor& input_,
+    const Tensor& grad_output_,
+    Tensor& grad_weight,
+    Tensor& grad_bias,
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    IntArrayRef output_padding,
+    IntArrayRef dilation,
+    int scale_) {
+  TORCH_CHECK(
+      kernel_size.size() == 3,
+      "It is expected kernel_size equals to 3, but got size ",
+      kernel_size.size());
+
+  TORCH_CHECK(
+      dilation.size() == 3,
+      "It is expected dilation equals to 3, but got size ",
+      dilation.size());
+
+  TORCH_CHECK(
+      padding.size() == 3,
+      "It is expected padding equals to 3, but got size ",
+      padding.size());
+
+  TORCH_CHECK(
+      stride.size() == 3,
+      "It is expected stride equals to 3, but got size ",
+      stride.size());
+
+  TORCH_CHECK(
+      output_padding.size() == 3,
+      "It is expected stride equals to 3, but got size ",
+      output_padding.size());
+
+  int64_t kernel_depth = kernel_size[0];
+  int64_t kernel_height = kernel_size[1];
+  int64_t kernel_width = kernel_size[2];
+  int64_t dilation_depth = dilation[0];
+  int64_t dilation_height = dilation[1];
+  int64_t dilation_width = dilation[2];
+  int64_t padding_depth = padding[0];
+  int64_t padding_height = padding[1];
+  int64_t padding_width = padding[2];
+  int64_t stride_depth = stride[0];
+  int64_t stride_height = stride[1];
+  int64_t stride_width = stride[2];
+  int64_t output_padding_depth = output_padding[0];
+  int64_t output_padding_height = output_padding[1];
+  int64_t output_padding_width = output_padding[2];
+
+  // number of input & output planes and kernel size is indirectly defined by
+  // the grad_weight tensor
+  slow_conv_transpose3d_shape_check(
```

Comment:
It looks like you factored out the shape check into that other header file, but I also see a bunch of other shape checking here. What's the split for (or is some of it duplicated)?

Reply:
Hey good question! Sorry I updated the descriptions of some PRs to explain the split, but missed this one. Essentially, I want to register a CPU dispatch, but the `REGISTER_DISPATCH` macro for CPU kernels requires that the code be place into the `native/cpu` dir where is it recompiled it once per arch type. So I moved all the backward logic underneath `native/cpu` so I can call `REGISTER_DISPATCH`.

I've been talking with Richard about this, and he is rightfully concerned that new multiple compilation unnecessarily regresses build time and expands the binary size. I've been throwing around an idea of defining a new macro `REGISTER_ALL_CPU_DISPATCH` that registers the same kernel across all arch types to avoid both the recompilation and the need to split the logic as done here. Do you hav any thoughts on this idea?